### DISCUSSION TITLE
求助，下面是我家云风扇调速脚本，因为我看不懂想请大佬按照我的要求修改一下增加一些功能。

### DISCUSSION THREAD

**Discussion by liyiranlab** [ophub/amlogic-s9xxx-armbian#2275] [Rockchip]
20240708 修复硬盘未待机低温切换未生效bug，今天以前的固件都有比较严重bug，用这些固件的请自己更新最下面脚本；
               调整未待机时的风扇启动停温度;
               修改部分无需变动的变量为未常量;

**Comment by liyiranlab** [comment]
另外，我不知道如何重启风扇脚本，请大佬告知，还有不知道我的要求能不能实现，谢谢大佬。

**Comment by liyiranlab** [comment]
通过曾经大二的c语言基础+GPT3.5 帮助，我完成了目标任务的代码添加。现在贴出来请大佬指点，测试了一下好像功能都正常工作了。

**Comment by liyiranlab** [comment]
```
#!/usr/bin/perl

use strict;

################################################################################
# 参数我已经修改完善，没有特殊情况勿改
# cat /sys/devices/virtual/thermal/thermal_zone0/temp #用此行命令可在ssh命令矿中查看cpu实时温度
# 参数调整区 风扇按照固定占空比启动后10s，根据cpu温度调整转速，低于$temp_low风扇停转，高于$temp_fanOn风扇启动。
# 速度最小值(满速是99),如果太小可能进入死区，风扇不转，需配合下面提示调整
my $speed_min = 8;

# 速度最大值(满速是99)
my $speed_max = 99;

# 温度低限(摄氏度): 小于等于此温度风扇停转
my $temp_low = 45;

# 风扇启动温度低限(摄氏度): 大于于等于此值按启动风扇散热
my $temp_fanOn = 65;

# 风扇启动占空比参数(%): 风扇启动时使用固定占空比：风扇启动占空比参数，设置要点:保证能启动同时风扇噪音不要过大
my $duty_cycle_on = 3;

# 温度最高限(摄氏度): 大于此值按最高速率转动
my $temp_high = 75;

# 调速间隔(秒)
my $interval = 10;
################################################################################
#用到的变量，下面参数勿动
# 风扇状态位: 0为关闭，1为开启
my $fanclose_temp = 0;

# 风扇首次启动: 0为风扇运转中，1风扇从关闭到开启是第一次运行
my $fan_up = 0;

# 风扇切换温度转换值
my $coeff_temp = 0;

################################################################################

my $fixed_speed = $ARGV[0];
my $period = 25000;
&init;
if( ($fixed_speed ne "") && 
    ($fixed_speed =~ m/^[0-9]{1,3}$/) ) {
    # 如果命令行参数为 0-100 的整数,则按指定的固定速率调速
    &set_fixed_speed($fixed_speed);
} else { 
    # 否则自动调速
    while(1) {
        &auto_speed;
        sleep($interval);    
    }
}
exit 0;

###############################################################################
sub get_soc_temp {
    my @fnames=(
          '/sys/devices/virtual/thermal/thermal_zone0/temp',
          '/sys/devices/platform/scpi/scpi:sensors/hwmon/hwmon0/temp1_input',
       );
    my $fh;
    my $temp = 50;
    for my $fname (@fnames) {
        if( -f $fname ) {
            open $fh, "<", $fname;
            $temp = <$fh> / 1000.0;
            close $fh;
            return $temp;
        }
    }
    return $temp;
}

sub init {

    $speed_max = 100  if ($speed_max > 100);
    $speed_min = 0  if($speed_min < 0);

    my $fh;
    `rmmod pwm_fan 2>/dev/null`;

    open $fh, ">", "/sys/class/pwm/pwmchip0/export";
    print $fh "0\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/period";
    print $fh "$period\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/polarity";
    print $fh "normal\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/enable";
    print $fh "1\n";
    close $fh;
}

sub set_fixed_speed {
    my $fixed_speed = shift;

    # 最大值0.99
    my $coeff_speed = ($fixed_speed / 100.0) > 1 ?  0.99 : ($fixed_speed / 100.0);

    #0.1-0.19 定义为死区
    $coeff_speed = 0.20 if $coeff_speed > 0 and $coeff_speed < 0.2;

    my $duty_cycle = int($coeff_speed * $period);

    open my $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/duty_cycle";
    print $fh "$duty_cycle\n";
    close $fh;
}

sub auto_speed {
    my $temp = &get_soc_temp;
    #my $coeff_temp;
    if($temp <= $temp_low) {
	$coeff_temp = 0;
	$fanclose_temp = 0;
	#$fanclose_temp = 1; #去掉此行代码前的#测试风扇是否可以维持旋转，测试完毕此行代码前加#
        } elsif($temp > $temp_high) {
                $coeff_temp = 1;
	        $fanclose_temp = 1;
                } elsif(($temp > $temp_fanOn)&&($fanclose_temp == 0)){
	             $fan_up = 1;
	             $fanclose_temp = 1;
	             }else {
                            $coeff_temp = ($temp - $temp_low) / ($temp_high - $temp_low);
                            }

    my $coeff_speed_min = $speed_min / $speed_max  * ($speed_max / 100);
    my $coeff_speed = $coeff_temp; 
    $coeff_speed = ($coeff_speed_min + $coeff_speed) > 0.99 ? 0.99 : ($coeff_speed_min + $coeff_speed);
   
    my $duty_cycle = int($coeff_speed * $period);

    if($fanclose_temp == 0) {
	$duty_cycle = 0;
        }elsif(($fanclose_temp == 1)&&($fan_up == 1)){
		$duty_cycle = $duty_cycle_on ;
		$fan_up = 0; #风扇以启动，清零首次标志位
		}else{
		}
    
    open my $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/duty_cycle";
    print $fh "$duty_cycle\n";
    close $fh;
}

```

**Comment by liyiranlab** [comment]
[pwm-fan.zip](https://github.com/ophub/amlogic-s9xxx-armbian/files/15161446/pwm-fan.zip)

**Comment by liyiranlab** [comment]
20240504 增加风扇降噪系数

**Comment by liyiranlab** [comment]
```
#!/usr/bin/perl

use strict;

################################################################################
# 参数我已经修改完善，没有特殊情况勿改
# cat /sys/devices/virtual/thermal/thermal_zone0/temp #用此行命令可在ssh命令矿中查看cpu实时温度
# 参数调整区 风扇按照固定占空比启动后10s,根据cpu温度调整转速,低于$temp_low风扇停转,高于$temp_fanOn风扇启动。
# 速度最小值(满速是99),如果太小可能进入死区,风扇不转,需配合下面提示调整
my $speed_min = 8;

# 速度最大值(满速是99)
my $speed_max = 99;

# 温度低限(摄氏度): 小于等于此温度风扇停转
my $temp_low = 45;

# 风扇启动温度低限(摄氏度): 大于于等于此值按启动风扇散热
my $temp_fanOn = 65;

# 风扇启动占空比参数(%): 风扇启动时使用固定占空比：风扇启动占空比参数,设置要点:保证能启动同时风扇噪音不要过大
my $duty_cycle_on = 3;

# 降噪系数: 降低风扇启动到cpu温度最高限区间的风扇噪音,降低幅度计算1/$fan_n,最终风扇转速=计算转速*(1/$fan_n)+$speed_min；
my $fan_n = 2;

# 温度最高限(摄氏度): 大于此值按最高速率转动
my $temp_high = 75;

# 调速间隔(秒)
my $interval = 10;
################################################################################
#用到的变量,下面参数勿动
# 风扇状态位:0为关闭,1为开启
my $fanclose_temp = 0;

# 风扇首次启动:0为风扇运转中,1风扇从关闭到开启是第一次运行
my $fan_up = 0;

# 风扇切换温度转换值
my $coeff_temp = 0;

################################################################################

my $fixed_speed = $ARGV[0];
my $period = 25000;
&init;
if( ($fixed_speed ne "") && 
    ($fixed_speed =~ m/^[0-9]{1,3}$/) ) {
    # 如果命令行参数为 0-100 的整数,则按指定的固定速率调速
    &set_fixed_speed($fixed_speed);
} else { 
    # 否则自动调速
    while(1) {
        &auto_speed;
        sleep($interval);    
    }
}
exit 0;

###############################################################################
sub get_soc_temp {
    my @fnames=(
          '/sys/devices/virtual/thermal/thermal_zone0/temp',
          '/sys/devices/platform/scpi/scpi:sensors/hwmon/hwmon0/temp1_input',
       );
    my $fh;
    my $temp = 50;
    for my $fname (@fnames) {
        if( -f $fname ) {
            open $fh, "<", $fname;
            $temp = <$fh> / 1000.0;
            close $fh;
            return $temp;
        }
    }
    return $temp;
}

sub init {

    $speed_max = 100  if ($speed_max > 100);
    $speed_min = 0  if($speed_min < 0);

    my $fh;
    `rmmod pwm_fan 2>/dev/null`;

    open $fh, ">", "/sys/class/pwm/pwmchip0/export";
    print $fh "0\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/period";
    print $fh "$period\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/polarity";
    print $fh "normal\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/enable";
    print $fh "1\n";
    close $fh;
}

sub set_fixed_speed {
    my $fixed_speed = shift;

    # 最大值0.99
    my $coeff_speed = ($fixed_speed / 100.0) > 1 ?  0.99 :($fixed_speed / 100.0);

    #0.1-0.19 定义为死区
    $coeff_speed = 0.20 if $coeff_speed > 0 and $coeff_speed < 0.2;

    my $duty_cycle = int($coeff_speed * $period);

    open my $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/duty_cycle";
    print $fh "$duty_cycle\n";
    close $fh;
}

sub auto_speed {
    my $temp = &get_soc_temp;
    #my $coeff_temp;
    if($temp <= $temp_low) {
	$coeff_temp = 0;
	$fanclose_temp = 0;
	#$fanclose_temp = 1; #去掉此行代码前的#测试风扇是否可以维持旋转,测试完毕此行代码前加#
        } elsif($temp > $temp_high) {
                $coeff_temp = 1;
	        $fanclose_temp = 1;
                } elsif(($temp > $temp_fanOn)&&($fanclose_temp == 0)){
	             $fan_up = 1; #风扇启转,设置首启转志位
	             $fanclose_temp = 1;
	             }else {
                            $coeff_temp = ($temp - $temp_low) / (($temp_high - $temp_low)*$fan_n);
                            }

    my $coeff_speed_min = $speed_min / $speed_max  * ($speed_max / 100);
    my $coeff_speed = $coeff_temp; 
    $coeff_speed = ($coeff_speed_min + $coeff_speed) > 0.99 ? 0.99 :($coeff_speed_min + $coeff_speed);
   
    my $duty_cycle = int($coeff_speed * $period);

    if($fanclose_temp == 0) {
	$duty_cycle = 0;
        }elsif(($fanclose_temp == 1)&&($fan_up == 1)){
		$duty_cycle = $duty_cycle_on ;
		$fan_up = 0; #风扇已启动,清零首启转志位
		}else{
		}
    
    open my $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/duty_cycle";
    print $fh "$duty_cycle\n";
    close $fh;
}

```

**Comment by liyiranlab** [comment]
[pwm-fan.zip](https://github.com/ophub/amlogic-s9xxx-armbian/files/15211585/pwm-fan.zip)

**Reply by ophub** [reply]
你这个脚本测试好了吗？允许我使用你的这个脚本添加到 https://github.com/ophub/amlogic-s9xxx-armbian/tree/main/build-armbian/armbian-files/different-files/chainedbox/rootfs/usr/bin 替换现在的文件可以吧。

你可以pull到这个位置，脚本里加个你的版权信息。

**Reply by liyiranlab** [reply]
还没用过github这些功能，小白我研究一下，最下面的代码目前没有遇到问题，各个功能我也分别测试了，应该没有问题。

**Reply by liyiranlab** [reply]
> 你这个脚本测试好了吗？允许我使用你的这个脚本添加到 https://github.com/ophub/amlogic-s9xxx-armbian/tree/main/build-armbian/armbian-files/different-files/chainedbox/rootfs/usr/bin 替换现在的文件可以吧。
> 
> 你可以pull到这个位置，脚本里加个你的版权信息。
大佬我只要fork你的项目，修改提交就可以了吧？

**Comment by liyiranlab** [comment]
替换/usr/bin/pwm-fan.pl

**Comment by liyiranlab** [comment]
降噪后的效果，降低噪音，风扇旋转时吹出热气，提高风扇散热效率。

**Comment by liyiranlab** [comment]
#!/usr/bin/perl

use strict;

################################################################################
#20240604 增加风扇降噪系数（系数可以是小数），倍数降低风扇温度最高限（最高限默认75度）以下，最低限以上的风扇转速，强调不超温就尽量不让风扇吵闹的原则；
#              修改风扇停转温度，根据硬盘最佳寿命40度为修改参考；
#20240428 增加风扇关闭功能，在冬天气温比较低时，如果风扇还旋转，因为热胀冷缩轴承非常容易磨损，天气冷的时候是风扇损坏率较高的时候。
# 参数我已经修改完善，没有特殊情况勿改 -根据F大脚本修改 by Ran
# cat /sys/devices/virtual/thermal/thermal_zone0/temp #用此行命令可在ssh命令矿中查看cpu实时温度
# 参数调整区 风扇按照固定占空比启动后10s,根据cpu温度调整转速,低于$temp_low风扇停转,高于$temp_fanOn风扇启动。
# 速度最小值(满速是99),如果太小可能进入死区,风扇不转,需配合下面提示调整
my $speed_min = 8;

# 速度最大值(满速是99)
my $speed_max = 99;

# 温度低限(摄氏度): 小于等于此温度风扇停转
my $temp_low = 40;

# 风扇启动温度低限(摄氏度): 大于于等于此值按启动风扇散热
my $temp_fanOn = 65;

# 风扇启动占空比参数(%): 风扇启动时使用固定占空比：风扇启动占空比参数,设置要点:保证能启动同时风扇噪音不要过大
my $duty_cycle_on = 3;

# 降噪系数: 降低风扇启动到cpu温度最高限区间的风扇噪音,降低幅度计算1/$fan_n,最终风扇转速=计算转速*(1/$fan_n)+$speed_min；
my $fan_n = 2;

# 温度最高限(摄氏度): 大于此值按最高速率转动
my $temp_high = 75;

# 调速间隔(秒)
my $interval = 10;
################################################################################
#用到的变量,下面参数勿动
# 风扇状态位:0为关闭,1为开启
my $fanclose_temp = 0;

# 风扇首次启动:0为风扇运转中,1风扇从关闭到开启是第一次运行
my $fan_up = 0;

# 风扇切换温度转换值
my $coeff_temp = 0;

################################################################################

my $fixed_speed = $ARGV[0];
my $period = 25000;
&init;
if( ($fixed_speed ne "") && 
    ($fixed_speed =~ m/^[0-9]{1,3}$/) ) {
    # 如果命令行参数为 0-100 的整数,则按指定的固定速率调速
    &set_fixed_speed($fixed_speed);
} else { 
    # 否则自动调速
    while(1) {
        &auto_speed;
        sleep($interval);    
    }
}
exit 0;

###############################################################################
sub get_soc_temp {
    my @fnames=(
          '/sys/devices/virtual/thermal/thermal_zone0/temp',
          '/sys/devices/platform/scpi/scpi:sensors/hwmon/hwmon0/temp1_input',
       );
    my $fh;
    my $temp = 50;
    for my $fname (@fnames) {
        if( -f $fname ) {
            open $fh, "<", $fname;
            $temp = <$fh> / 1000.0;
            close $fh;
            return $temp;
        }
    }
    return $temp;
}

sub init {

    $speed_max = 100  if ($speed_max > 100);
    $speed_min = 0  if($speed_min < 0);

    my $fh;
    `rmmod pwm_fan 2>/dev/null`;

    open $fh, ">", "/sys/class/pwm/pwmchip0/export";
    print $fh "0\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/period";
    print $fh "$period\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/polarity";
    print $fh "normal\n";
    close $fh;

    open $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/enable";
    print $fh "1\n";
    close $fh;
}

sub set_fixed_speed {
    my $fixed_speed = shift;

    # 最大值0.99
    my $coeff_speed = ($fixed_speed / 100.0) > 1 ?  0.99 :($fixed_speed / 100.0);

    #0.1-0.19 定义为死区
    $coeff_speed = 0.20 if $coeff_speed > 0 and $coeff_speed < 0.2;

    my $duty_cycle = int($coeff_speed * $period);

    open my $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/duty_cycle";
    print $fh "$duty_cycle\n";
    close $fh;
}

sub auto_speed {
    my $temp = &get_soc_temp;
    #my $coeff_temp;
    if($temp <= $temp_low) {
	$coeff_temp = 0;
	$fanclose_temp = 0;
	#$fanclose_temp = 1; #去掉此行代码前的#测试风扇是否可以维持旋转,测试完毕此行代码前加#
        } elsif($temp > $temp_high) {
                $coeff_temp = 1;
	              $fanclose_temp = 1;
                } elsif(($temp > $temp_fanOn)&&($fanclose_temp == 0)){
	                       $fan_up = 1; #风扇启转,设置首启转志位
	                       $fanclose_temp = 1;
	                       }else {
                            $coeff_temp = ($temp - $temp_low) / (($temp_high - $temp_low)*$fan_n);
                            }

    my $coeff_speed_min = $speed_min / $speed_max  * ($speed_max / 100);
    my $coeff_speed = $coeff_temp; 
    $coeff_speed = ($coeff_speed_min + $coeff_speed) > 0.99 ? 0.99 :($coeff_speed_min + $coeff_speed);
   
    my $duty_cycle = int($coeff_speed * $period);

    if($fanclose_temp == 0) {
	$duty_cycle = 0;
        }elsif(($fanclose_temp == 1)&&($fan_up == 1)){
		$duty_cycle = $duty_cycle_on ;
		$fan_up = 0; #风扇已启动,清零首启转志位
		}else{
		}
    
    open my $fh, ">", "/sys/class/pwm/pwmchip0/pwm0/duty_cycle";
    print $fh "$duty_cycle\n";
    close $fh;
}

**Comment by liyiranlab** [comment]
[pwm-fan.zip](https://github.com/user-attachments/files/15554032/pwm-fan.zip)

**Reply by liyiranlab** [reply]
最新版本

**Comment by liyiranlab** [comment]
[2024-07-08-fixpwm-fan.zip](https://github.com/user-attachments/files/16122658/2024-07-08-fixpwm-fan.zip)
最新版本比较完善。

**Reply by liyiranlab** [reply]
> [2024-07-08-fixpwm-fan.zip](https://github.com/user-attachments/files/16122658/2024-07-08-fixpwm-fan.zip) 最新版本比较完善。
20240708 修复硬盘未待机低温切换未生效bug，今天以前的固件都有比较严重bug，用这些固件的请自己更新；
               调整未待机时的风扇启动停温度;
               修改部分无需变动的变量为未常量;

**Reply by zcmi** [reply]
你好，这个脚本在我家云的openwrt中可以使用吗？

**Comment by tv4you2016** [comment]
hello @liyiranlab , 
Is this script compatible with the HK1 box?
dtb used: [meson-sm1-hk1box-vontar-x3.dtb](https://github.com/ophub/amlogic-s9xxx-armbian/blob/cc97ec0f276c337f839401322e5f2ebd7fed3830/build-armbian/armbian-files/platform-files/amlogic/bootfs/dtb/amlogic/meson-sm1-hk1box-vontar-x3.dtb)

**Reply by liyiranlab** [reply]
这个脚本是给我家云的，如果HK1 box 能用之前我家云原始的版本，那么就可以用这个修改版。

**Comment by jackdanny77** [comment]
各位大神脚本有了，问题是具体怎么使用，能出一下教学吗

**Reply by liyiranlab** [reply]
https://www.right.com.cn/forum/forum.php?mod=viewthread&tid=8373013&page=1&mobile=2#pid20969343

**Comment by liyiranlab** [comment]
[https://github.com/ophub/amlogic-s9xxx-armbian/tree/main/build-armbian/armbian-files/different-files/chainedbox/rootfs/usr/bin](url)
这里是源码中最新版本。
